nanomq/pub_handler_test.c

```diff
@@ -0,0 +1,105 @@
+#include <assert.h>
+#include <stdio.h>
+
+#include "include/nanomq.h"
+#include "include/pub_handler.h"
+
+int
+main()
+{
+	int         rv      = -1;
+	reason_code rv_rc   = -1;
+	bool        rv_bool = false;
+	// init work
+	nano_work *work;
+	work            = nng_alloc(sizeof(*work));
+	work->proto_ver = MQTT_PROTOCOL_VERSION_v311;
+
+	nng_msg *msg;
+	nng_msg_alloc(&msg, 0);
+
+	struct pub_packet_struct *pub_packet;
+	pub_packet                  = nng_zalloc(sizeof(*pub_packet));
+
+	work->msg                   = msg;
+	work->pub_packet            = pub_packet;
+
+	struct fixed_header *fix_hd = nng_alloc(sizeof(*fix_hd));
+	fix_hd->qos                 = 0;
+	fix_hd->packet_type         = PUBLISH;
+
+	// test data
+	uint32_t remaining_len      = 18;
+	uint32_t topic_len          = 7;
+	uint32_t data_len           = 4;
+	// topic: $MQTT
+	uint8_t topic[] = {
+		0x00, 0x05 /* topic length */, 0x24, 0x4D, 0x51, 0x54,
+		0x54 /* topic body*/
+	};
+	// data: data
+	uint8_t data[] = { 0x64, 0x61, 0x74, 0x61 };
+
+	// init msg->body and msg->header
+	nng_msg_append(work->msg, topic, topic_len);
+	nng_msg_append(work->msg, data, data_len);
+	nng_msg_set_remaining_len(work->msg, remaining_len);
+	nng_msg_header_append(work->msg, fix_hd, sizeof(*fix_hd));
+
+	/* test for decode_pub_message */
+	// TODO test for MQTTv5
+	// test for remaining_len > msg_len
+	rv_rc = decode_pub_message(work, MQTT_PROTOCOL_VERSION_v311);
+	assert(rv_rc == PROTOCOL_ERROR);
+	// test for commom case.
+	remaining_len = 11;
+	nng_msg_set_remaining_len(work->msg, remaining_len);
+	rv_rc = decode_pub_message(work, MQTT_PROTOCOL_VERSION_v311);
+	assert(rv_rc == SUCCESS);
+	// check work->pub_packet
+	assert(work->pub_packet->var_header.publish.topic_name.len == 5);
+	assert(strcmp(work->pub_packet->var_header.publish.topic_name.body,
+	           "$MQTT") == 0);
+	assert(work->pub_packet->payload.len == 4);
+	assert(strcmp(pub_packet->payload.data, "data") == 0);
```

Comment:
Payload is formatted as binary. It's better to compare bytes in payload one by one.